packages/dd-trace/src/appsec/api_security_sampler.js

```diff
@@ -1,61 +1,87 @@
 'use strict'
 
+const TTLCache = require('@isaacs/ttlcache')
+const web = require('../plugins/util/web')
 const log = require('../log')
+const { AUTO_REJECT, USER_REJECT } = require('../../../../ext/priority')
 
-let enabled
-let requestSampling
+const MAX_SIZE = 4096
+const DEFAULT_DELAY = 30 // 30s
 
-const sampledRequests = new WeakSet()
+let enabled
+let sampledRequests
 
 function configure ({ apiSecurity }) {
   enabled = apiSecurity.enabled
-  setRequestSampling(apiSecurity.requestSampling)
+  const delay = apiSecurity.sampleDelay || DEFAULT_DELAY
+  sampledRequests = new TTLCache({ max: MAX_SIZE, ttl: delay * 1000 })
 }
 
 function disable () {
   enabled = false
+  sampledRequests?.clear()
 }
 
-function setRequestSampling (sampling) {
-  requestSampling = parseRequestSampling(sampling)
-}
+function sampleRequest (req, res, force = false) {
+  if (!enabled) return false
+  if (isSampled(req, res)) return false
+
+  const rootSpan = web.root(req)
+  if (!rootSpan) return false
 
-function parseRequestSampling (requestSampling) {
-  let parsed = parseFloat(requestSampling)
+  let priority = getSpanPriority(rootSpan)
+  if (!priority) {
+    rootSpan._prioritySampler?.sample(rootSpan)
+    priority = getSpanPriority(rootSpan)
+  }
 
-  if (isNaN(parsed)) {
-    log.warn(`Incorrect API Security request sampling value: ${requestSampling}`)
+  if (priority === AUTO_REJECT || priority === USER_REJECT) {
+    return false
+  }
 
-    parsed = 0
-  } else {
-    parsed = Math.min(1, Math.max(0, parsed))
+  if (force) {
+    return sample(req, res)
```

Comment:
I think you could replace this whole line by just

```suggestion
    sampledRequests.set(key)
```
I think it would be equivalent

Reply:
why was this comment resolved @IlyasShabi ? when resolving comments it's nice to put a conclusion in the thread so we know why we dismissed it ?

Reply:
@simon-id I resolved this issue based on your proposal. I removed the `sample()` function and replaced it with your suggested approach, you can notice that the original function is now outdated.

Reply:
I think what @simon-id means is that there is a check at the beginning of the sampleRequest function to know if the req is sampled so if you have reached this point there is no need to check again if `sampledRequests.has(key)`. The above check ensures that it is not.

Reply:
Yes it's clear now, I did push a fix